aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,919 @@
+            }) {
+                return TestResult::failed();
+            }
+        }
+        TestResult::passed()
+    }
+
+    #[quickcheck]
+    fn hpke_key_rotator_remove_config(
+        config: HpkeKeyRotatorConfig,
+        state: InitialGlobalHpkeKeysState,
+    ) -> TestResult {
+        let clock = MockClock::new(state.start);
+
+        let to_expire: HashSet<_> = state
+            .keypairs
+            .iter()
+            .filter(|(_, keypair)| {
+                !config.ciphersuites.contains(&keypair.ciphersuite())
+                    && keypair.state() == &HpkeKeyState::Active
+            })
+            .map(|(id, _)| *id)
+            .collect();
+
+        let to_delete: HashSet<_> = state
+            .keypairs
+            .iter()
+            .filter(|(_, keypair)| {
+                !config.ciphersuites.contains(&keypair.ciphersuite())
+                    && keypair.state() == &HpkeKeyState::Pending
+            })
+            .map(|(id, _)| *id)
+            .collect();
+
+        if to_expire.is_empty() && to_delete.is_empty() {
+            return TestResult::discard();
+        }
+
+        let key_rotator = HpkeKeyRotator::new(clock, state.keypairs, &config)
+            .unwrap()
+            .sweep()
+            .unwrap();
+
+        if to_delete
+            .iter()
+            .any(|id| key_rotator.keypairs.contains_key(id))
+        {
+            TestResult::error("pending key should have been deleted")
+        } else if !to_expire
+            .iter()
+            .all(|id| key_rotator.keypairs.get(id).unwrap().state() == &HpkeKeyState::Expired)
+        {
+            TestResult::error("active key should have been expired")
+        } else {
+            TestResult::passed()
+        }
+    }
+
+    #[quickcheck]
+    fn hpke_key_rotator_new_config(
+        config: HpkeKeyRotatorConfig,
+        state: InitialGlobalHpkeKeysState,
+    ) -> TestResult {
+        if state.keypairs.is_empty() {
+            return TestResult::discard();
+        }
+        let clock = MockClock::new(state.start);
+
+        let ciphersuites_to_insert: Vec<_> = config
+            .ciphersuites
+            .iter()
+            .filter(|&ciphersuite| {
+                !state
+                    .keypairs
+                    .iter()
+                    .any(|(_, keypair)| &keypair.ciphersuite() == ciphersuite)
+            })
+            .collect();
+
+        let key_rotator = HpkeKeyRotator::new(clock, state.keypairs, &config)
+            .unwrap()
+            .sweep()
+            .unwrap();
+
+        for ciphersuite in ciphersuites_to_insert {
+            let inserted: Vec<_> = key_rotator
+                .keypairs
+                .iter()
+                .filter(|(_, keypair)| &keypair.ciphersuite() == ciphersuite)
+                .collect();
+            if inserted.len() != 1 {
+                return TestResult::error("more than one key inserted for new ciphersuite");
```

Comment:
Clarify why? `TestResult::error()` looks the same as `TestResult::failed()` except it sets a friendly error message https://docs.rs/quickcheck/latest/src/quickcheck/tester.rs.html#216. That way if a test fails it points us directly to the failing line, rather than having to run with RUST_LOG.

Reply:
In both `error()` and `failed()` cases, we're presented with the input variables, only in `error()` we're given an additional line `Error: expired key was not deleted`.

e.g.
```
# cargo test hpke_key_rotator_expired_keys
   Compiling janus_core v0.7.19 (/home/inahga/Projects/janus/core)
   Compiling janus_aggregator_core v0.7.19 (/home/inahga/Projects/janus/aggregator_core)
   Compiling janus_aggregator_api v0.7.19 (/home/inahga/Projects/janus/aggregator_api)
   Compiling janus_aggregator v0.7.19 (/home/inahga/Projects/janus/aggregator)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 58.37s
     Running unittests src/lib.rs (/home/inahga/Projects/janus/target/debug/deps/janus_aggregator-4c9e31edbd742197)

running 1 test
test aggregator::key_rotator::tests::hpke_key_rotator_expired_keys ... FAILED

failures:

---- aggregator::key_rotator::tests::hpke_key_rotator_expired_keys stdout ----
thread 'aggregator::key_rotator::tests::hpke_key_rotator_expired_keys' panicked at /home/inahga/.cargo/registry/src/index.crates.io-6f17d22bba15001f/quickcheck-1.0.3/src/tester.rs:165:28:
[quickcheck] TEST FAILED (runtime error). Arguments: [the arguments, really long, somewhat justifying fixing `Arbitrary::shrink`.]
Error: expired key was not deleted
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    aggregator::key_rotator::tests::hpke_key_rotator_expired_keys

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 193 filtered out; finished in 0.05s
```